tensorflow/python/distribute/collective_all_reduce_strategy.py

```diff
@@ -880,7 +880,9 @@ def experimental_should_init(self):
 
   @property
   def should_checkpoint(self):
-    return self._is_chief
+    #return self._is_chief
+    task_type = ['chief', 'worker', 'evaluator']
+    return True if self._task_type in task_type else False
```

Comment:
What's reason to checkpoint on non-chief workers?

Reply:
Because if non-chief workers don’t restore the model from the checkpoint, training can‘t continue.

For example, I am using `tf.estimator.Estimator API` and `tf.distribute.CollectiveAllReduceStrategy` for distributed training.

Firstly, I train for 1000 rounds(train_stpes=1000) and save the checkpoint, it works normally.

Then I set the train_steps to 2000, only the is_chief role can restore the model from the checkpoint without any error.

Chief worker output as follow:
```
INFO:tensorflow:Graph was finalized.
I1218 15:24:28.721170 139824828032832 monitored_session.py:240] Graph was finalized.
INFO:tensorflow:Restoring parameters from /tmp/iris/iris-chief-0/checkpoint/model.ckpt-1000
I1218 15:24:28.722573 139824828032832 saver.py:1284] Restoring parameters from /tmp/iris/iris-chief-0/checkpoint/model.ckpt-1000
WARNING:tensorflow:From /root/PycharmProjects/multi_gpu_demo/venv/lib/python3.6/site-packages/tensorflow_core/python/training/saver.py:1069: get_checkpoint_mtimes (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
Instructions for updating:
Use standard file utilities to get mtimes.
W1218 15:24:28.753480 139824828032832 deprecation.py:323] From /root/PycharmProjects/multi_gpu_demo/venv/lib/python3.6/site-packages/tensorflow_core/python/training/saver.py:1069: get_checkpoint_mtimes (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
Instructions for updating:
Use standard file utilities to get mtimes.
INFO:tensorflow:Running local_init_op.
I1218 15:24:28.782264 139824828032832 session_manager.py:500] Running local_init_op.
INFO:tensorflow:Done running local_init_op.
I1218 15:24:28.938534 139824828032832 session_manager.py:502] Done running local_init_op.
INFO:tensorflow:Saving checkpoints for 1000 into /tmp/iris/iris-chief-0/checkpoint/model.ckpt.
I1218 15:24:29.215120 139824828032832 basic_session_run_hooks.py:606] Saving checkpoints for 1000 into /tmp/iris/iris-chief-0/checkpoint/model.ckpt.
```
Non-chief worker output as follow:

```
INFO:tensorflow:Graph was finalized.
I1218 15:24:59.388847 140629246773056 monitored_session.py:240] Graph was finalized.
```

https://github.com/tensorflow/tensorflow/blob/b1f506e1ea81de05d83c679c02b58cabb696260c/tensorflow/python/distribute/collective_all_reduce_strategy.py#L881-L883

When I change non-chief worker to should-checkpoit，the training can continue at the last checkpoint.

I think,  in AllReduce mode, each node only saves part of the model parameters. When loading the model from the 

checkpoint, it also needs to be loaded from the checkpoint of each node

Reply:
https://github.com/tensorflow/tensorflow/blob/41802b3785a3e89f421254de6b6326de516cdb8d/tensorflow/python/training/monitored_session.py#L404-L419

When should_checkpoint is true, checkpoint save path is as follows:

```
root@05246d8eb041:/home/tf# ll /tmp/iris/iris-chief-0/checkpoint/
total 2496
drwxr-xr-x 2 root root   4096 Dec 23 02:26 ./
drwxr-xr-x 4 root root   4096 Dec 23 02:24 ../
-rw-r--r-- 1 root root    132 Dec 23 02:26 checkpoint
-rw-r--r-- 1 root root 804320 Dec 23 02:24 events.out.tfevents.1608690276.05246d8eb041
-rw-r--r-- 1 root root 772096 Dec 23 02:26 events.out.tfevents.1608690360.05246d8eb041
-rw-r--r-- 1 root root 501291 Dec 23 02:26 graph.pbtxt
-rw-r--r-- 1 root root   1560 Dec 23 02:26 model.ckpt-1000.data-00000-of-00001
-rw-r--r-- 1 root root    661 Dec 23 02:26 model.ckpt-1000.index
-rw-r--r-- 1 root root 220943 Dec 23 02:26 model.ckpt-1000.meta
-rw-r--r-- 1 root root   1560 Dec 23 02:24 model.ckpt-500.data-00000-of-00001
-rw-r--r-- 1 root root    661 Dec 23 02:24 model.ckpt-500.index
-rw-r--r-- 1 root root 220943 Dec 23 02:24 model.ckpt-500.meta
```

When should_checkpoint is false, checkpoint save path is as follows:
```
root@05246d8eb041:/home/tf# ll /tmp/iris/iris-worker-0/checkpoint/tmp_worker_0/
total 2036
drwxr-xr-x 2 root root   4096 Dec 23 02:26 ./
drwxr-xr-x 3 root root   4096 Dec 23 02:24 ../
-rw-r--r-- 1 root root    132 Dec 23 02:24 checkpoint
-rw-r--r-- 1 root root 786080 Dec 23 02:24 events.out.tfevents.1608690284.05246d8eb041
-rw-r--r-- 1 root root 337188 Dec 23 02:26 events.out.tfevents.1608690369.05246d8eb041
-rw-r--r-- 1 root root 491872 Dec 23 02:24 graph.pbtxt
-rw-r--r-- 1 root root   1560 Dec 23 02:24 model.ckpt-1000.data-00000-of-00001
-rw-r--r-- 1 root root    661 Dec 23 02:24 model.ckpt-1000.index
-rw-r--r-- 1 root root 216638 Dec 23 02:24 model.ckpt-1000.meta
-rw-r--r-- 1 root root   1560 Dec 23 02:24 model.ckpt-500.data-00000-of-00001
-rw-r--r-- 1 root root    661 Dec 23 02:24 model.ckpt-500.index
-rw-r--r-- 1 root root 216638 Dec 23 02:24 model.ckpt-500.meta
```
When should_checkpoint is false, a temporary path(such as tmp_worker_0) will be generated and the checkpoint will be saved in it.

If I move the files in the temporary path to the parent directory, the training can continue.

`mv /tmp/iris/iris-worker-0/checkpoint/tmp_worker_0/* /tmp/iris/iris-worker-0/checkpoint/`

Reply:
@crccw  Can you please take a look on the above comment from @UrmsOne. Thanks!

Reply:
It appears that you're using local paths on each worker, i.e. non-chief workers cannot read the checkpoint written by the chief worker. The existing logic has the assumption that non-chief workers are able to read the checkpoint written by the chief worker.

The expected logic is that if should_checkpoint is true, write checkpoint to the path; otherwise write to a temp path and delete it. The reason the latter is needed is because checkpointing may trigger cross device communication, so all workers must checkpoint, and to a different path.

This unfortunately fails if you use local path. Is it possible that you use a network path? We cannot change the definition of should_checkpoint as it's backward incompatible. You're welcome to change estimator to accommodate this case.

Reply:
@UrmsOne Can you please check @crccw's comments and keep us posted ? Thanks!

Reply:
@gbaned Sorry for my delayed reply. As @crccw  said, there is no problem if use NFS, but my datasets are all above 1TB, using NFS will brings additional storage overhead.  So I suggest that a switch should be reserved, when using local storage,  the checkpoint file can be written by themselves. This also needs to consider the issue of backward compatibility.